synfig-studio/src/gui/workarearenderer/renderer_canvas.cpp

```diff
@@ -371,19 +376,75 @@ Renderer_Canvas::build_onion_frames()
-			if (time >= rend_desc.get_time_start() && time <= rend_desc.get_time_end())
-				onion_frames.push_back(FrameDesc(time, w, h, alpha));
+
+		if (keyframes)
+		{
+			// Onion Skin on Keyframes
+			// Find and save past Keyframes
+			Time keyframe_time = base_time;
+			for(int i = past; i > 0; --i) {
+				KeyframeList::iterator iter;
+				if (canvas->keyframe_list().find_prev(keyframe_time, iter)) {
+					if (keyframe_time >= rend_desc.get_time_start() && keyframe_time <= rend_desc.get_time_end()) {
+						keyframe_time = iter->get_time();
+						past_keyframes.push_back(keyframe_time);
+					}
+				}
+				else
+					break;
+			}
+
+			// Find and save future Keyframes
+			keyframe_time = base_time;
+			for(int i = future; i > 0; --i) {
+				KeyframeList::iterator iter;
+				if (canvas->keyframe_list().find_next(keyframe_time, iter)) {
+					if (keyframe_time >= rend_desc.get_time_start() && keyframe_time <= rend_desc.get_time_end()) {
+						keyframe_time = iter->get_time();
+						future_keyframes.push_back(keyframe_time);
+					}
+				}
+				else
+					break;
+			}
+
+			// Check number of available past and future Keyframes
+			int past_frames = (past > past_keyframes.size()) ? past_keyframes.size() : past;
+			int future_frames = (future > future_keyframes.size()) ? future_keyframes.size() : future;
+
+			// Cycle through past Keyframes and retrieve their Time values
+			for(int i = past_frames; i > 0; --i) {
+				ColorReal alpha = base_alpha + (ColorReal)(past_frames - i + 1)/(ColorReal)(past_frames + 1);
+				onion_frames.push_back(FrameDesc(past_keyframes[past_frames - i], w, h, alpha));
+			}
+
+			// Cycle through future Keyframes and retrieve their Time values
+			for(int i = future_frames; i > 0; --i) {
+				ColorReal alpha = base_alpha + (ColorReal)(future_frames - i + 1)/(ColorReal)(future_frames + 1);
+				onion_frames.push_back(FrameDesc(future_keyframes[future_frames - i], w, h, alpha));
+			}
 		}
-		for(int i = future; i > 0; --i) {
-			Time time = base_time + frame_duration*i;
-			ColorReal alpha = base_alpha + (ColorReal)(future - i + 1)/(ColorReal)(future + 1);
-			if (time >= rend_desc.get_time_start() && time <= rend_desc.get_time_end())
-				onion_frames.push_back(FrameDesc(time, w, h, alpha));
+		else
+		{
+			// Onion Skin on Frames
+			// Past Frames
+			for(int i = past; i > 0; --i) {
+				Time time = base_time - frame_duration*i;
+				ColorReal alpha = base_alpha + (ColorReal)(past - i + 1)/(ColorReal)(past + 1);
+				if (time >= rend_desc.get_time_start() && time <= rend_desc.get_time_end())
+					onion_frames.push_back(FrameDesc(time, w, h, alpha));
+			}
+
+			// Future Frames
+			for(int i = future; i > 0; --i) {
+				Time time = base_time + frame_duration*i;
+				ColorReal alpha = base_alpha + (ColorReal)(future - i + 1)/(ColorReal)(future + 1);
+				if (time >= rend_desc.get_time_start() && time <= rend_desc.get_time_end())
+					onion_frames.push_back(FrameDesc(time, w, h, alpha));
+			}
```

Comment:
One last thing about this part:

You can see the two codes here are too similar here, right?
It's good to avoid such 'repetition' to prevent someone to change something in one code (fixing it somehow) and forget to check if second code should be changed too.

I'd suggest to populate the vectors for both past and future onion skin times no matter if it is set to keyframes or frames. Next, you loop like you did on lines 415-425.

Something like:
```c++
std::vector<Time> past_times;
std::vector<Time> future_times;

if (is_keyframes) {
   // loop for filling past_times and future_times according to valid keyframes
} else {
   // loop for filling past_times and future_times according to fps and frames
}

for(int i = past_frames; i > 0; --i) {
	ColorReal alpha = base_alpha + (ColorReal)(past_frames - i + 1)/(ColorReal)(past_frames + 1);
	onion_frames.push_back(FrameDesc(past_times[past_frames - i], w, h, alpha));
}

for(int i = future_frames; i > 0; --i) {
	ColorReal alpha = base_alpha + (ColorReal)(future_frames - i + 1)/(ColorReal)(future_frames + 1);
	onion_frames.push_back(FrameDesc(future_times[future_frames - i], w, h, alpha));
}
```

Reply:
Done :+1: